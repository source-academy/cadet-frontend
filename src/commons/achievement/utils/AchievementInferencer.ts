import assert from 'assert';

import {
  AchievementGoal,
  AchievementItem,
  AchievementStatus,
  FilterStatus
} from '../../../features/achievement/AchievementTypes';

/**
 * An InferencerNode item encapsulates all important information of an achievement item
 *
 * @param {AchievementItem} achievement the achievement item
 * @param {number} dataIdx the key to retrive the achivement item in achievements[], i.e. achievements[dataIdx] = achievement
 * @param {number} maxExp total achievable EXP of the achievement
 * @param {number} progressFrac progress percentage in fraction. It is always between 0 to 1, both inclusive.
 * @param {AchievementStatus} status the achievement status
 * @param {Date | undefined} displayDeadline deadline displayed on the achievement card
 * @param {Set<number>} children a set of immediate prerequisites id
 * @param {Set<number>} descendant a set of all descendant prerequisites id (including immediate prerequisites)
 */
class InferencerNode {
  public achievement: AchievementItem;
  public dataIdx: number;
  public maxExp: number;
  public progressFrac: number;
  public status: AchievementStatus;
  public displayDeadline?: Date;
  public children: Set<number>;
  public descendant: Set<number>;

  constructor(achievement: AchievementItem, dataIdx: number) {
    const { deadline, prerequisiteIds, goals } = achievement;

    this.achievement = achievement;
    this.dataIdx = dataIdx;
    this.maxExp = this.generateMaxExp(goals);
    this.progressFrac = this.generateProgressFrac(goals);
    this.status = AchievementStatus.ACTIVE; // to be updated after the nodeList is constructed
    this.displayDeadline = deadline;
    this.children = new Set(prerequisiteIds);
    this.descendant = new Set(prerequisiteIds);
  }

  private generateMaxExp(goals: AchievementGoal[]) {
    return goals.reduce((exp, goal) => exp + goal.goalTarget, 0);
  }

  private generateProgressFrac(goals: AchievementGoal[]) {
    const progress = goals.reduce((progress, goal) => progress + goal.goalProgress, 0);

    return Math.min(progress / this.maxExp, 1);
  }
}

/**
 * Main Class which handles the assignment of ids to each and every Achievement in the system.
 */
class AchievementInferencer {
  private achievements: AchievementItem[] = []; // note: the achievement_id might not be the same as its array index
  private nodeList: Map<number, InferencerNode> = new Map(); // key = achievement_id, value = achievement node

  constructor(achievements: AchievementItem[]) {
    this.achievements = achievements;
    this.processData();
  }

  public doesAchievementExist(id: number) {
    return this.nodeList.get(id) !== undefined;
  }

  public getAchievements() {
    return this.achievements;
  }

  public getAchievementItem(id: number) {
    assert(this.nodeList.has(id));
    return this.nodeList.get(id)!.achievement;
  }

  public insertAchievement(achievement: AchievementItem) {
    // first, generate a new unique id by finding the max id
    let newId = 0;
    if (this.achievements.length > 0) {
      newId = Math.max(...this.nodeList.keys(), 0) + 1;
    }

    // then assign the new unique id by overwriting the achievement item supplied by param
    // and append it to achievements[]
    achievement.id = newId;
    this.achievements.push(achievement);

    // finally, reconstruct the nodeList
    this.processData();

    return newId;
  }

  public modifyAchievement(achievement: AchievementItem) {
    // directly modify the achievement element in achievements
    assert(this.nodeList.has(achievement.id));
    const idx = this.nodeList.get(achievement.id)!.dataIdx;
    this.achievements[idx] = achievement;

    // then, reconstruct the nodeList
    this.processData();
  }

  public removeAchievement(id: number) {
    const hasChild = (achievement: AchievementItem) => achievement.prerequisiteIds.includes(id);

    const removeChild = (achievement: AchievementItem) =>
      achievement.prerequisiteIds.filter(child => child !== id);

    // create a copy of achievements that:
    // 1. does not contain the removed achievement
    // 2. does not contain reference of the removed achievement in other achievement's prerequisite
    const newAchievements: AchievementItem[] = [];
    this.achievements.forEach(achievement => {
      if (hasChild(achievement)) {
        // reference of the removed item is filtered out
        achievement.prerequisiteIds = removeChild(achievement);
      }
      if (achievement.id !== id) {
        // removed achievement is not included in the new achievements
        newAchievements.push(achievement);
      }
    });
    this.achievements = newAchievements;

    // finally, reconstruct the nodeList
    this.processData();
  }

  public listIds() {
    return this.achievements.map(achievement => achievement.id);
  }

  public listTaskIds() {
    return this.achievements.filter(achievement => achievement.isTask).map(task => task.id);
  }

  public listTaskIdsbyPosition() {
    return this.achievements
      .filter(achievement => achievement.isTask)
      .sort((taskA, taskB) => taskA.position - taskB.position)
      .map(task => task.id);
  }

  public listNonTaskIds() {
    return this.achievements.filter(achievement => !achievement.isTask).map(nonTask => nonTask.id);
  }

  public setTask(achievement: AchievementItem) {
    achievement.isTask = true;
    achievement.position = this.listTaskIds().length;

    this.modifyAchievement(achievement);
    this.normalizePositions();
  }

  public setNonTask(achievement: AchievementItem) {
    achievement.prerequisiteIds = [];
    achievement.isTask = false;
    achievement.position = 0; // position 0 is reserved for non-task achievements

    this.modifyAchievement(achievement);
    this.normalizePositions();
  }

  // Calculates set bonus
  public getBonusExp(id: number) {
    assert(this.nodeList.has(id));
    if (this.nodeList.get(id)!.children.size === 0) return 0;

    const maxExp = this.nodeList.get(id)!.maxExp;

    let maxChildExp = 0;
    for (const childId of this.nodeList.get(id)!.children) {
      maxChildExp = maxChildExp + this.nodeList.get(childId)!.maxExp;
    }

    return maxExp - maxChildExp;
  }

  public getMaxExp(id: number) {
    assert(this.nodeList.has(id));
    return this.nodeList.get(id)!.maxExp;
  }

  public getStudentExp(id: number) {
    assert(this.nodeList.has(id));
    const goals = this.nodeList.get(id)!.achievement.goals;

    return goals.reduce((progress, goal) => progress + goal.goalProgress, 0);
  }

  // total achievable EXP of all published achievements
  public getTotalExp() {
    const publishedTask = this.listPublishedNodes().filter(node => node.achievement.isTask);

    return publishedTask.reduce((totalExp, node) => totalExp + node.maxExp, 0);
  }

  // total EXP earned by the student
  public getStudentTotalExp() {
    const publishedTask = this.listPublishedNodes().filter(node => node.achievement.isTask);

    return publishedTask.reduce((totalProgress, node) => {
      const goals = node.achievement.goals;
      const progress = goals.reduce((progress, goal) => progress + goal.goalProgress, 0);
      return totalProgress + progress;
    }, 0);
  }

  public getProgressFrac(id: number) {
    assert(this.nodeList.has(id));
    return this.nodeList.get(id)!.progressFrac;
  }

  public getStatus(id: number) {
    assert(this.nodeList.has(id));
    return this.nodeList.get(id)!.status;
  }

  public getDisplayDeadline(id: number) {
    assert(this.nodeList.has(id));
    return this.nodeList.get(id)!.displayDeadline;
  }

  public isImmediateChild(id: number, childId: number) {
    assert(this.nodeList.has(id));
    return this.nodeList.get(id)!.children.has(childId);
  }

  public getImmediateChildren(id: number) {
    assert(this.nodeList.has(id));
    return this.nodeList.get(id)!.children;
  }

  public listImmediateChildren(id: number) {
    return [...this.getImmediateChildren(id)];
  }

  public isDescendant(id: number, childId: number) {
    assert(this.nodeList.has(id));
    return this.nodeList.get(id)!.descendant.has(childId);
  }

  public getDescendants(id: number) {
    assert(this.nodeList.has(id));
    return this.nodeList.get(id)!.descendant;
  }

  public listDescendants(id: number) {
    return [...this.getDescendants(id)];
  }

  public listAvailablePrerequisites(id: number) {
    return this.listIds().filter(
      target => target !== id && !this.isDescendant(id, target) && !this.isDescendant(target, id)
    );
  }

  public getFilterCount(filterStatus: FilterStatus) {
    const published = this.listPublishedNodes();

    switch (filterStatus) {
      case FilterStatus.ALL:
        return published.length;
      case FilterStatus.ACTIVE:
        return published.filter(node => node.status === AchievementStatus.ACTIVE).length;
      case FilterStatus.COMPLETED:
        return published.filter(node => node.status === AchievementStatus.COMPLETED).length;
      default:
        return 0;
    }
  }

  // NOTE: positions of achievements are 1-indexed.
  public changeAchievementPosition(achievement: AchievementItem, newPosition: number) {
    const achievements = this.getAchievements()
      .filter(achievement => achievement.isTask)
      .sort((taskA, taskB) => taskA.position - taskB.position);

    const movedAchievement = achievements.splice(achievement.position - 1, 1)[0];
    achievements.splice(newPosition - 1, 0, movedAchievement);

    for (let i = Math.min(newPosition - 1, achievement.position); i < achievements.length; i++) {
      const editedAchievement = achievements[i];
      editedAchievement.position = i + 1;
    }
  }

  private processData() {
    this.constructNodeList();
    this.nodeList.forEach(node => {
      this.generateDescendant(node);
      this.generateDisplayDeadline(node);
      this.generateStatus(node);
    });
  }

  private constructNodeList() {
    this.nodeList = new Map();
    for (let idx = 0; idx < this.achievements.length; idx++) {
      const achievement = this.achievements[idx];
      this.nodeList.set(achievement.id, new InferencerNode(achievement, idx));
    }
  }

  // Recursively append grandchildren's id to children, O(N) operation
  private generateDescendant(node: InferencerNode) {
    for (const childId of node.descendant) {
      if (childId === node.achievement.id) {
        console.error('Circular dependency detected');
      }
      assert(this.nodeList.has(childId));
      for (const grandchildId of this.nodeList.get(childId)!.descendant) {
        // Newly added grandchild is appended to the back of the set.
        node.descendant.add(grandchildId);
        // Hence the great grandchildren will be added when the iterator reaches there
      }
    }
  }

  // Set the node's display deadline by comparing with all descendants' deadlines
  private generateDisplayDeadline(node: InferencerNode) {
    const now = new Date();

    // Comparator of two deadlines
    const compareDeadlines = (
      displayDeadline: Date | undefined,
      currentDeadline: Date | undefined
    ) => {
      if (currentDeadline === undefined || currentDeadline <= now) {
        // currentDeadline undefined or expired, nothing change
        return displayDeadline;
      } else if (displayDeadline === undefined) {
        return currentDeadline;
      } else {
        // currentDeadline unexpired, displayDeadline may be expired or unexpired
        // display the closest unexpired deadline
        return displayDeadline <= now || currentDeadline < displayDeadline
          ? currentDeadline
          : displayDeadline;
      }
    };

    // Temporary array of all descendants' deadlines
    const descendantDeadlines = [];
    for (const childId of node.descendant) {
      assert(this.nodeList.has(childId));
      const childDeadline = this.nodeList.get(childId)!.achievement.deadline;
      descendantDeadlines.push(childDeadline);
    }

    // Reduces the temporary array to a single Date value
    node.displayDeadline = descendantDeadlines.reduce(compareDeadlines, node.displayDeadline);
  }

  private generateStatus(node: InferencerNode) {
    const now = new Date();
    const deadline = node.displayDeadline;
    if (deadline !== undefined && deadline.getTime() <= now.getTime()) {
      // deadline elapsed
      if (node.progressFrac === 0) {
        return (node.status = AchievementStatus.EXPIRED); // not attempted
      } else {
        return (node.status = AchievementStatus.COMPLETED); // attempted
      }
    } else {
      // deadline not elapsed
      if (node.progressFrac === 1) {
        return (node.status = AchievementStatus.COMPLETED); // fully completed
      } else {
        return (node.status = AchievementStatus.ACTIVE); // not fully completed
      }
    }
  }

  // normalize positions
  private normalizePositions() {
    this.achievements.sort((a, b) => a.position - b.position);

    // position 0 is reserved for non-task achievements
    const nonTaskPosition = 0;
    let newPosition = 1;

    for (let idx = 0; idx < this.achievements.length; idx++) {
      if (this.achievements[idx].isTask) {
        this.achievements[idx].position = newPosition++;
      } else {
        this.achievements[idx].position = nonTaskPosition;
      }
    }
  }

  private listPublishedNodes() {
    // returns an array of Node that are published to the achievement page
    return this.listTaskIds().reduce((arr, id) => {
      assert(this.nodeList.has(id));
      const node = this.nodeList.get(id)!;
      arr.push(node); // including task achievement
      for (const child of node.children) {
        assert(this.nodeList.has(child));
        arr.push(this.nodeList.get(child)!); // including immediate prerequisites
      }
      return arr;
    }, [] as InferencerNode[]);
  }
}

export default AchievementInferencer;
