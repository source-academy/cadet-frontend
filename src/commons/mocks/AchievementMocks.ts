import {
  AchievementAbility,
  AchievementGoal,
  AchievementItem,
  GoalType
} from '../../features/achievement/AchievementTypes';

export const mockAchievements: AchievementItem[] = [
  {
    id: 0,
    title: 'Rune Master',
    ability: AchievementAbility.CORE,
    isTask: true,
    position: 1,
    prerequisiteIds: [2, 1],
    goalIds: [0],
    cardTileUrl:
      'https://source-academy-assets.s3-ap-southeast-1.amazonaws.com/achievement/card-tile/rune-master-tile.png',
    view: {
      canvasUrl:
        'https://source-academy-assets.s3-ap-southeast-1.amazonaws.com/achievement/canvas/sample-canvas.png',
      description:
        'Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud',
      completionText: 'Congratulations! You have completed Rune Master achievement'
    }
  },
  {
    id: 1,
    title: 'Beyond the Second Dimension',
    ability: AchievementAbility.CORE,
    deadline: new Date(2020, 7, 6, 12, 30, 0),
    isTask: false,
    position: 0,
    prerequisiteIds: [],
    goalIds: [1],
    cardTileUrl:
      'https://source-academy-assets.s3-ap-southeast-1.amazonaws.com/achievement/card-tile/btsd-tile.png',
    view: {
      canvasUrl:
        'https://source-academy-assets.s3-ap-southeast-1.amazonaws.com/achievement/canvas/sample-canvas.png',
      description:
        'Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud',
      completionText: 'Congratulations! You have completed Beyond the Second Dimension achievement'
    }
  },
  {
    id: 2,
    title: 'Colorful Carpet',
    ability: AchievementAbility.CORE,
    deadline: new Date(2020, 7, 8, 9, 0, 0),
    isTask: false,
    position: 0,
    prerequisiteIds: [],
    goalIds: [2],
    cardTileUrl:
      'https://source-academy-assets.s3-ap-southeast-1.amazonaws.com/achievement/card-tile/colorful-carpet-tile.png',
    view: {
      canvasUrl:
        'https://source-academy-assets.s3-ap-southeast-1.amazonaws.com/achievement/canvas/sample-canvas.png',
      description:
        'Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud',
      completionText: 'Congratulations! You have completed Colorful Carpet achievement'
    }
  },
  {
    id: 3,
    title: '',
    ability: AchievementAbility.CORE,
    isTask: false,
    position: 0,
    prerequisiteIds: [],
    goalIds: [],
    cardTileUrl: '',
    view: {
      canvasUrl: '',
      description: '',
      completionText: ''
    }
  },
  {
    id: 4,
    title: 'Curve Wizard',
    ability: AchievementAbility.CORE,
    deadline: new Date(2020, 8, 15, 0, 0, 0),
    isTask: true,
    position: 4,
    prerequisiteIds: [5, 6],
    goalIds: [3],
    cardTileUrl:
      'https://source-academy-assets.s3-ap-southeast-1.amazonaws.com/achievement/card-tile/curve-wizard-tile.png',
    view: {
      canvasUrl:
        'https://source-academy-assets.s3-ap-southeast-1.amazonaws.com/achievement/canvas/sample-canvas.png',
      description:
        'Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud',
      completionText: 'Congratulations! You have completed Curve Wizard achievement'
    }
  },
  {
    id: 5,
    title: 'Curve Introduction',
    ability: AchievementAbility.CORE,
    deadline: new Date(2020, 7, 28, 0, 0, 0),
    isTask: false,
    position: 0,
    prerequisiteIds: [],
    goalIds: [4],
    cardTileUrl:
      'https://source-academy-assets.s3-ap-southeast-1.amazonaws.com/achievement/card-tile/curve-introduction-tile.png',
    view: {
      canvasUrl:
        'https://source-academy-assets.s3-ap-southeast-1.amazonaws.com/achievement/canvas/sample-canvas.png',
      description:
        'Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud',
      completionText: 'Congratulations! You have completed Curve Introduction achievement'
    }
  },
  {
    id: 6,
    title: 'Curve Manipulation',
    ability: AchievementAbility.CORE,
    deadline: new Date(2020, 8, 5, 0, 0, 0),
    isTask: false,
    position: 0,
    prerequisiteIds: [],
    goalIds: [5],
    cardTileUrl:
      'https://source-academy-assets.s3-ap-southeast-1.amazonaws.com/achievement/card-tile/curve-manipulation-tile.png',
    view: {
      canvasUrl:
        'https://source-academy-assets.s3-ap-southeast-1.amazonaws.com/achievement/canvas/sample-canvas.png',
      description:
        'Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud',
      completionText: 'Congratulations! You have completed Curve Manipulation achievement'
    }
  },
  {
    id: 7,
    title: 'The Source-rer',
    ability: AchievementAbility.EFFORT,
    deadline: new Date(2020, 7, 21, 0, 0, 0),
    isTask: true,
    position: 3,
    prerequisiteIds: [],
    goalIds: [6, 7],
    cardTileUrl:
      'https://source-academy-assets.s3-ap-southeast-1.amazonaws.com/achievement/card-tile/the-source-rer-tile.png',
    view: {
      canvasUrl:
        'https://source-academy-assets.s3-ap-southeast-1.amazonaws.com/achievement/canvas/sample-canvas.png',
      description:
        'Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud',
      completionText: 'Congratulations! You have completed The Source-rer achievement'
    }
  },
  {
    id: 8,
    title: 'Power of Friendship',
    ability: AchievementAbility.COMMUNITY,
    isTask: true,
    position: 2,
    prerequisiteIds: [9],
    goalIds: [],
    cardTileUrl:
      'https://source-academy-assets.s3-ap-southeast-1.amazonaws.com/achievement/card-tile/power-of-friendship-tile.png',
    view: {
      canvasUrl:
        'https://source-academy-assets.s3-ap-southeast-1.amazonaws.com/achievement/canvas/sample-canvas.png',
      description:
        'Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud',
      completionText: 'Congratulations! You have completed Power of Friendship achievement'
    }
  },
  {
    id: 9,
    title: 'Piazza Guru',
    ability: AchievementAbility.COMMUNITY,
    isTask: false,
    position: 0,
    prerequisiteIds: [],
    goalIds: [8],
    cardTileUrl:
      'https://source-academy-assets.s3-ap-southeast-1.amazonaws.com/achievement/card-tile/piazza-guru-tile.png',
    view: {
      canvasUrl:
        'https://source-academy-assets.s3-ap-southeast-1.amazonaws.com/achievement/canvas/sample-canvas.png',
      description:
        'Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud',
      completionText: 'Congratulations! You have completed Piazza Guru achievement'
    }
  },
  {
    id: 10,
    title: "That's the Spirit",
    ability: AchievementAbility.EXPLORATION,
    isTask: true,
    position: 5,
    prerequisiteIds: [],
    goalIds: [9],
    cardTileUrl:
      'https://source-academy-assets.s3-ap-southeast-1.amazonaws.com/achievement/card-tile/thats-the-spirit-tile.png',
    view: {
      canvasUrl:
        'https://source-academy-assets.s3-ap-southeast-1.amazonaws.com/achievement/canvas/sample-canvas.png',
      description:
        'Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud',
      completionText: "Congratulations! You have completed That's the Spirit achievement"
    }
  },
  {
    id: 11,
    title: 'Kool Kidz',
    ability: AchievementAbility.FLEX,
    isTask: true,
    position: 6,
    prerequisiteIds: [],
    goalIds: [],
    cardTileUrl:
      'https://source-academy-assets.s3-ap-southeast-1.amazonaws.com/achievement/card-tile/annotated-tile.png',
    view: {
      canvasUrl:
        'https://source-academy-assets.s3-ap-southeast-1.amazonaws.com/achievement/canvas/sample-canvas.png',
      description: 'Diz for teh K00L K1dz',
      completionText: 'U SO G00D'
    }
  }
];

export const mockGoals: AchievementGoal[] = [
  {
    id: 0,
    text: 'Bonus for completing Colorful Carpet & Beyond the Second Dimension Achievements',
    maxExp: 100,
    type: GoalType.BINARY,
    meta: {
      condition: `AND(
        {
          event: 'achievement',
          restriction: 2,
        },
        {
          event: 'achievement',
          restriction: 1,
        },
      )`
    },
    exp: 0,
    completed: false
  },
  {
    id: 1,
    text: 'XP earned from Beyond the Second Dimension Mission',
    maxExp: 250,
    type: GoalType.ASSESSMENT,
    meta: { assessmentId: 5, requiredCompletionExp: 200 },
    exp: 213,
    completed: true
  },
  {
    id: 2,
    text: 'XP earned from Colorful Carpet Mission',
    maxExp: 250,
    type: GoalType.ASSESSMENT,
    meta: { assessmentId: 3, requiredCompletionExp: 200 },
    exp: 0,
    completed: false
  },
  {
    id: 3,
    text: 'Bonus for completing Curve Introduction & Curve Manipulation Achievements',
    maxExp: 100,
    type: GoalType.BINARY,
    meta: {
      condition: `AND(
      {
        event: 'achievement',
        restriction: 4,
      },
      {
        event: 'achievement',
        restriction: 5,
      },
    )`
    },
    exp: 0,
    completed: false
  },
  {
    id: 4,
    text: 'XP earned from Curve Introduction Mission',
    maxExp: 250,
    type: GoalType.ASSESSMENT,
    meta: { assessmentId: 7, requiredCompletionExp: 150 },
    exp: 178,
    completed: true
  },
  {
    id: 5,
    text: 'XP earned from Curve Manipulation Mission',
    maxExp: 250,
    type: GoalType.ASSESSMENT,
    meta: { assessmentId: 8, requiredCompletionExp: 200 },
    exp: 191,
    completed: false
  },
  {
    id: 6,
    text: 'Submit Source 3 Path',
    maxExp: 100,
    type: GoalType.BINARY,
    meta: {
      condition: `
      {
        event: 'assessment-submission',
        restriction: paths.SOURCE_THREE,
      },
    `
    },
    exp: 100,
    completed: true
  },
  {
    id: 7,
    text: 'XP earned from Source 3 Path',
    maxExp: 300,
    type: GoalType.ASSESSMENT,
    meta: {
      assessmentId: 12,
      requiredCompletionExp: 300
    },
    exp: 300,
    completed: true
  },
  {
    id: 8,
    text: 'Each Top Voted answer in Piazza gives 10 XP',
    maxExp: 100,
    type: GoalType.MANUAL,
    meta: {},
    exp: 40,
    completed: false
  },
  {
    id: 9,
    text: 'Submit 1 PR to Source Academy Github',
    maxExp: 100,
    type: GoalType.MANUAL,
    meta: {},
    exp: 100,
    completed: true
  },
  {
    id: 10,
    text: 'Be the Koolest Kidz in SOC by redeeming this 100 XP achievement yourself',
    maxExp: 100,
    type: GoalType.BINARY,
    meta: {
      condition: `
      {
        event: 'achievement',
        restriction: 'false',
      },
    `
    },
    exp: 0,
    completed: false
  }
];
