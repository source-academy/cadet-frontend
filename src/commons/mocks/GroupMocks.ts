import { GradingSummary } from '../../features/dashboard/DashboardTypes';

export const mockGradingSummary: GradingSummary = [
  {
    leaderName: 'John',
    groupName: 'Mock Group 1',
    ungradedMissions: 123,
    submittedMissions: 200,
    ungradedSidequests: 100,
    submittedSidequests: 117
  },
  {
    leaderName: 'Molly',
    groupName: 'Mock Group 2',
    ungradedMissions: 1232,
    submittedMissions: 205430,
    ungradedSidequests: 345,
    submittedSidequests: 11547
  },
  {
    leaderName: 'Lenny',
    groupName: 'Mock Group 3',
    ungradedMissions: 1532,
    submittedMissions: 22200,
    ungradedSidequests: 134500,
    submittedSidequests: 6777
  }
];
