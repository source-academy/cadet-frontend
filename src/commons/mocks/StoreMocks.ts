import { Store } from 'redux';
import mockStore from 'redux-mock-store';

import {
  defaultAcademy,
  defaultApplication,
  defaultDashboard,
  defaultPlayground,
  defaultSession,
  defaultWorkspaceManager,
  OverallState
} from '../application/ApplicationTypes';

export function mockInitialStore<P>(): Store<OverallState> {
  const createStore = (mockStore as any)();
  const state: OverallState = {
    academy: defaultAcademy,
    application: defaultApplication,
    dashboard: defaultDashboard,
    playground: defaultPlayground,
    workspaces: defaultWorkspaceManager,
    session: defaultSession
  };
  return createStore(state);
}
