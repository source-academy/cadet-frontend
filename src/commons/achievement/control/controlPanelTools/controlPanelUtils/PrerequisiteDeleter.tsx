import { Button } from '@blueprintjs/core';
import React, { useState } from 'react';

import { AchievementItem } from '../../../../../features/achievement/AchievementTypes';
import AchievementInferencer from '../../../utils/AchievementInferencer';
import AchievementSelection from './AchievementSelection';

type PrerequisiteDeleterProps = {
  editableAchievement: AchievementItem;
  setEditableAchievement: any;
  inferencer: AchievementInferencer;
  saveChanges: any;
};

function PrerequisiteDeleter(props: PrerequisiteDeleterProps) {
  const { editableAchievement, setEditableAchievement, inferencer, saveChanges } = props;

  const [isDialogOpen, setDialogOpen] = useState<boolean>(false);
  const toggleDialogOpen = () => setDialogOpen(!isDialogOpen);

  const prerequisites = editableAchievement.prerequisiteIds;

  const deletePrerequisite = (prerequisiteID: number) => {
    const newAchievement = editableAchievement;
    newAchievement.prerequisiteIds.filter(id => id !== prerequisiteID);
    return newAchievement;
  };

  const [deletedPrerequisiteID, setDeletedPrerequisiteID] = useState<number>(
    prerequisites.length === 0 ? 0 : prerequisites[0]
  );

  const deleteAction = () => {
    setEditableAchievement(deletePrerequisite(deletedPrerequisiteID));
    inferencer.modifyAchievement(editableAchievement);

    saveChanges();

    toggleDialogOpen();
  };

  return (
    <>
      <Button className="editor-button" onClick={toggleDialogOpen} text={'Delete A Prerequisite'} />
      <AchievementSelection
        selections={prerequisites}
        inferencer={inferencer}
        selectedId={deletedPrerequisiteID}
        setSelectedId={setDeletedPrerequisiteID}
        buttonText={'Delete Prerequisite'}
        dialogHeader={'Delete A Prerequisite'}
        emptySelectionsMessage={'You have no more prerequisites to delete'}
        toggleDialogOpen={toggleDialogOpen}
        isDialogOpen={isDialogOpen}
        action={deleteAction}
      />
    </>
  );
}

export default PrerequisiteDeleter;
