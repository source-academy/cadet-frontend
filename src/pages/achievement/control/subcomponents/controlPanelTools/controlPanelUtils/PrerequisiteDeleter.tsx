import { Button } from '@blueprintjs/core';
import React, { useState } from 'react';
import AchievementInferencer from 'src/pages/achievement/dashboard/subcomponents/utils/AchievementInferencer';

import { AchievementItem } from '../../../../../../commons/achievement/AchievementTypes';
import AchievementSelector from './AchievementSelector';

type PrerequisiteDeleterProps = {
  editableAchievement: AchievementItem;
  setEditableAchievement: any;
  inferencer: AchievementInferencer;
  saveChanges: any;
};

function PrerequisiteDeleter(props: PrerequisiteDeleterProps) {
  const { editableAchievement, setEditableAchievement, inferencer, saveChanges } = props;

  const [isDialogOpen, setDialogOpen] = useState<boolean>(false);
  const toggleDialogOpen = () => setDialogOpen(!isDialogOpen);

  const prerequisites = editableAchievement.prerequisiteIds;

  const deletePrerequisite = (prerequisiteID: number) => {
    const newAchievement = editableAchievement;

    for (let i = 0; i < newAchievement.prerequisiteIds.length; i++) {
      if (newAchievement.prerequisiteIds[i] === prerequisiteID) {
        newAchievement.prerequisiteIds.splice(i, 1);
      }
    }

    return newAchievement;
  };

  const [deletedPrerequisiteID, setDeletedPrerequisiteID] = useState<number>(
    prerequisites.length === 0 ? 0 : prerequisites[0]
  );

  const deleteAction = () => {
    setEditableAchievement(deletePrerequisite(deletedPrerequisiteID));
    inferencer.modifyAchievement(editableAchievement);

    saveChanges();

    toggleDialogOpen();
  };

  return (
    <>
      <Button className="editor-button" onClick={toggleDialogOpen} text={'Delete A Prerequisite'} />
      <AchievementSelector
        selections={prerequisites}
        inferencer={inferencer}
        selectedId={deletedPrerequisiteID}
        setSelectedId={setDeletedPrerequisiteID}
        buttonText={'Delete Prerequisite'}
        dialogHeader={'Delete A Prerequisite'}
        emptySelectionsMessage={'You have no more prerequisites to delete'}
        toggleDialogOpen={toggleDialogOpen}
        isDialogOpen={isDialogOpen}
        action={deleteAction}
      />
    </>
  );
}

export default PrerequisiteDeleter;
