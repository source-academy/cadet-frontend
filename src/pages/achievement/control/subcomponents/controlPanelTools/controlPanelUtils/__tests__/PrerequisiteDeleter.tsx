import { mount } from 'enzyme';
import React from 'react';

import { mockAchievements } from '../../../../../../..//commons/mocks/AchievementMocks';
import AchievementInferencer from '../../../../../dashboard/subcomponents/utils/AchievementInferencer';
import PrerequisiteDeleter from '../PrerequisiteDeleter';

const mockProps = {
  editableAchievement: mockAchievements[0],
  setEditableAchievement: () => {},
  inferencer: new AchievementInferencer(mockAchievements),
  saveChanges: () => {}
};

test('PrerequisiteDeleter component renders correctly', () => {
  const component = <PrerequisiteDeleter {...mockProps} />;
  const tree = mount(component);
  expect(tree.debug()).toMatchSnapshot();
});
