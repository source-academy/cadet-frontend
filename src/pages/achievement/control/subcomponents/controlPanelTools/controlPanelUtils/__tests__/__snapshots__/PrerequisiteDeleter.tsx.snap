// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PrerequisiteDeleter component renders correctly 1`] = `
"<PrerequisiteDeleter editableAchievement={{...}} setEditableAchievement={[Function: setEditableAchievement]} inferencer={{...}} saveChanges={[Function: saveChanges]}>
  <Blueprint3.Button className=\\"editor-button\\" onClick={[Function: toggleDialogOpen]} text=\\"Delete A Prerequisite\\">
    <button type=\\"button\\" className=\\"bp3-button editor-button\\" onClick={[Function: toggleDialogOpen]} disabled={[undefined]} onKeyDown={[Function (anonymous)]} onKeyUp={[Function (anonymous)]} tabIndex={[undefined]}>
      <Blueprint3.Icon icon={[undefined]} />
      <span className=\\"bp3-button-text\\">
        Delete A Prerequisite
      </span>
      <Blueprint3.Icon icon={[undefined]} />
    </button>
  </Blueprint3.Button>
  <AchievementSelector selections={{...}} inferencer={{...}} selectedId={2} setSelectedId={[Function: bound dispatchAction]} buttonText=\\"Delete Prerequisite\\" dialogHeader=\\"Delete A Prerequisite\\" emptySelectionsMessage=\\"You have no more prerequisites to delete\\" toggleDialogOpen={[Function: toggleDialogOpen]} isDialogOpen={false} action={[Function: deleteAction]}>
    <Blueprint3.Dialog onClose={[Function: toggleDialogOpen]} className=\\"task-selector\\" isOpen={false} title=\\"Delete A Prerequisite\\" canOutsideClickClose={true}>
      <Blueprint3.Overlay onClose={[Function: toggleDialogOpen]} className=\\"bp3-overlay-scroll-container\\" isOpen={false} title=\\"Delete A Prerequisite\\" canOutsideClickClose={true} hasBackdrop={true} autoFocus={true} backdropProps={{...}} canEscapeKeyClose={true} enforceFocus={true} lazy={true} transitionDuration={300} transitionName=\\"bp3-overlay\\" usePortal={true} />
    </Blueprint3.Dialog>
  </AchievementSelector>
</PrerequisiteDeleter>"
`;
